src/transformers/modeling_bert.py

```diff
@@ -409,14 +412,25 @@ def forward(
             if self.output_hidden_states:
                 all_hidden_states = all_hidden_states + (hidden_states,)
 
-            layer_outputs = layer_module(
-                hidden_states,
-                attention_mask,
-                head_mask[i],
-                encoder_hidden_states,
-                encoder_attention_mask,
-                output_attentions,
-            )
+            if getattr(self.config, "gradient_checkpointing", False):
+                layer_outputs = torch.utils.checkpoint.checkpoint(
+                    layer_module,
+                    hidden_states,
+                    attention_mask,
+                    head_mask[i],
+                    encoder_hidden_states,
+                    encoder_attention_mask,
+                    output_attentions,
```

Comment:
This is not working. `torch.utils.checkpoint.checkpoint` only accepts objects of type `Variable`

Reply:
@LysandreJik, can you help with this one? A recent commit added the `output_attentions` boolean argument and this breaks `torch.utils.checkpoint.checkpoint` because it is expecting all its parameters to be of type `Variable`. One solution I tried is to replace 
```
layer_outputs = torch.utils.checkpoint.checkpoint(
                    layer_module,
                    hidden_states,
                    attention_mask,
                    head_mask[i],
                    encoder_hidden_states,
                    encoder_attention_mask,
                    output_attentions
                )   
```

with

```
def my_fn(*inputs):
    return layer_module(*inputs, output_attentions)
layer_outputs = torch.utils.checkpoint.checkpoint(
                    my_fn,
                    hidden_states,
                    attention_mask,
                    head_mask[i],
                    encoder_hidden_states,
                    encoder_attention_mask,
                ) 
```

but it breaks with DPP. Try 

```
import os
import torch
from transformers import BertForPreTraining
os.environ['MASTER_ADDR'] = 'localhost'
os.environ['MASTER_PORT'] = '12355'
torch.distributed.init_process_group(backend="nccl", rank=0, world_size=1)

model = BertForPreTraining.from_pretrained('bert-base-uncased').cuda()
model = torch.nn.parallel.DistributedDataParallel(model, find_unused_parameters=True)
outputs = model(torch.tensor([[1, 2, 3]]).cuda())
outputs[0].sum().backward()
```

and you will get the error

```
RuntimeError: has_marked_unused_parameters_ INTERNAL ASSERT FAILED at /tmp/pip-req-build-4baxydiv/torch/csrc/distributed/c10d/reducer.cpp:290, please report a bug to PyTorch. 
```

* Why is it breaking with DDP?

* Is there a different solution that doesn't break DDP?

* or, can we pass the argument `output_attentions` to the layer in a different way?

Reply:
@LysandreJik, fixed in this commit https://github.com/ibeltagy/transformers/commit/5eb68bb804f5ffbfc7ba13c45a47717f72d04574, but don't know why the commit didn't show up in the PR here.